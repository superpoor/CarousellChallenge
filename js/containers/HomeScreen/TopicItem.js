import React from 'react'
import {
  View, 
  Text,
  StyleSheet
} from 'react-native'
import PropTypes from 'prop-types'

class TopicItem extends React.PureComponent {
  static propTypes = {
    topic: PropTypes.object.isRequired
  };

  render() {
    const { topic } = this.props;

    return (
      <View style={styles.container}>
        <Text style={styles.content}>{topic.content}</Text>
        <View style={styles.voteContainer}>
          <View style={styles.voteCountContainer}>
            <Text style={styles.vote}>{topic.votes} votes</Text>
          </View>
        </View>
      </View>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    backgroundColor: 'white',
    marginVertical: 5,
    elevation: 3,
    paddingHorizontal: 20,
    paddingVertical: 15,
  },
  content: {
    color: '#545454',
    fontSize: 14
  },
  voteContainer: {
    flexDirection: 'row',
    marginTop: 10,
  },
  voteCountContainer: {
    backgroundColor: '#F6F6F6',
    borderRadius: 2,
    padding: 8,
  },
  vote: {
    color: '#747474',
  }
})

export default TopicItem;