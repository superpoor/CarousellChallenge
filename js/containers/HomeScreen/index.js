import React from 'react'
import {
  View, 
  Text,
  Button,
  FlatList,
  StyleSheet
} from 'react-native'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'

import TopicItem from './TopicItem'
import Loading from '../../components/loading'
import { fetchTopics, updateVote } from '../../actions/topics'
import { getSortedTopicsByVote } from '../../selectors/topics' 

export class HomeScreen extends React.PureComponent {

  static propTypes = {
    topics: PropTypes.array.isRequired,
    isFetchingTopics: PropTypes.bool.isRequired,
    isUpdatingVoteIds: PropTypes.array.isRequired,
    fetchTopics: PropTypes.func.isRequired,
    updateVote: PropTypes.func.isRequired
  };

  static navigationOptions = ({ navigation }) => {
    return {
      headerTitle: "Carousell",
      headerRight: (
        <Button  
          title="NEW"
          onPress={() => navigation.navigate("NewTopic")}
        />
      )
    }
  };

  _renderItem = ({item}) => {
    return (
      <TopicItem key={item.id} topic={item} isUpdatingVote={this.props.isUpdatingVoteIds.includes(item.id)} onUpdateVote={this._onUpdateVote} />
    )
  }

  render() {
    const { topics, isFetchingTopics, isUpdatingVoteIds } = this.props;

    if (isFetchingTopics) return <Loading />

    return (
      <View style={styles.listContainer}>
        <FlatList 
          keyExtractor={item => item.id}
          extraData={isUpdatingVoteIds}
          data={topics}
          renderItem={this._renderItem}
        />
      </View>
    )
  }

  componentDidMount() {
    this.props.fetchTopics();
  }

  _onUpdateVote = (topicId, deltaValue) => {
    this.props.updateVote(topicId, deltaValue);
  }
}


const styles = StyleSheet.create({
  listContainer: {
    flex: 1,
    backgroundColor: '#E2E2E2'
  }
})


export default connect(
  state => ({
    topics: getSortedTopicsByVote(state),
    isFetchingTopics: state.topics.isFetching,
    isUpdatingVoteIds: state.topics.isUpdatingVoteIds,
  }),
  dispatch => bindActionCreators({
    fetchTopics,
    updateVote
  }, dispatch)
)(HomeScreen);
