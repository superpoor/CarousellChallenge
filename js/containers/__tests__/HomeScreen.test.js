import React from 'react';
import renderer from 'react-test-renderer';

import { HomeScreen } from '../HomeScreen'

describe('Test HomeScreen component', () => {

  it('renders correct HomeScreen component with topics', () => {
    const rendered = renderer.create(
      <HomeScreen 
        topics={[{
          id: "123", "content": "content1", votes: 1
        }, {
          id: "124", "content": "content2", votes: 5
        }]}
        isFetchingTopics={false}
        isUpdatingVoteIds={[]}
        fetchTopics={() => {}}
        updateVote={() => {}}
      />
    ).toJSON();
    expect(rendered).toMatchSnapshot();
  })

  it('renders correct HomeScreen component while fetching topic', () => {
    const rendered = renderer.create(
      <HomeScreen 
        topics={[]}
        isFetchingTopics={true}
        isUpdatingVoteIds={[]}
        fetchTopics={() => {}}
        updateVote={() => {}}
      />
    ).toJSON();
    expect(rendered).toMatchSnapshot();
  })

  it('renders correct HomeScreen component while updating vote of some topics', () => {
    const rendered = renderer.create(
      <HomeScreen 
        topics={[{
          id: "123", "content": "content1", votes: 1
        }, {
          id: "124", "content": "content2", votes: 5
        }]}
        isFetchingTopics={false}
        isUpdatingVoteIds={["124"]}
        fetchTopics={() => {}}
        updateVote={() => {}}
      />
    ).toJSON();
    expect(rendered).toMatchSnapshot();
  })
})

