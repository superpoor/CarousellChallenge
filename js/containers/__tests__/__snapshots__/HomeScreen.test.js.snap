// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test HomeScreen component renders correct HomeScreen component while fetching topic 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <ActivityIndicator
    animating={true}
    color="#999999"
    hidesWhenStopped={true}
    size="large"
  />
</View>
`;

exports[`Test HomeScreen component renders correct HomeScreen component while updating vote of some topics 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#E2E2E2",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    data={
      Array [
        Object {
          "content": "content1",
          "id": "123",
          "votes": 1,
        },
        Object {
          "content": "content2",
          "id": "124",
          "votes": 5,
        },
      ]
    }
    disableVirtualization={false}
    extraData={
      Array [
        "124",
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "white",
              "elevation": 3,
              "marginVertical": 5,
              "paddingHorizontal": 20,
              "paddingVertical": 15,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#545454",
                "fontSize": 14,
              }
            }
          >
            content1
          </Text>
          <View
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "marginTop": 10,
              }
            }
          >
            <View
              style={
                Object {
                  "backgroundColor": "#F6F6F6",
                  "borderRadius": 2,
                  "padding": 8,
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Object {
                    "color": "#747474",
                  }
                }
              >
                1
                 votes
              </Text>
            </View>
            <View
              accessibilityComponentType={undefined}
              accessibilityLabel={undefined}
              accessibilityTraits={undefined}
              accessible={true}
              collapsable={undefined}
              hasTVPreferredFocus={undefined}
              hitSlop={undefined}
              isTVSelectable={true}
              nativeID={undefined}
              onLayout={undefined}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#B02623",
                  "borderRadius": 3,
                  "margin": 5,
                  "opacity": 1,
                  "paddingHorizontal": 20,
                  "paddingVertical": 10,
                }
              }
              testID={undefined}
              tvParallaxProperties={undefined}
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Array [
                    Object {
                      "fontSize": 16,
                      "fontWeight": "bold",
                    },
                    Object {
                      "color": "white",
                    },
                  ]
                }
              >
                UP VOTE
              </Text>
            </View>
            <View
              accessibilityComponentType={undefined}
              accessibilityLabel={undefined}
              accessibilityTraits={undefined}
              accessible={true}
              collapsable={undefined}
              hasTVPreferredFocus={undefined}
              hitSlop={undefined}
              isTVSelectable={true}
              nativeID={undefined}
              onLayout={undefined}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#D8D8D8",
                  "borderRadius": 3,
                  "margin": 5,
                  "opacity": 1,
                  "paddingHorizontal": 20,
                  "paddingVertical": 10,
                }
              }
              testID={undefined}
              tvParallaxProperties={undefined}
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Array [
                    Object {
                      "fontSize": 16,
                      "fontWeight": "bold",
                    },
                    Object {
                      "color": "#000000",
                    },
                  ]
                }
              >
                DOWN VOTE
              </Text>
            </View>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "white",
              "elevation": 3,
              "marginVertical": 5,
              "paddingHorizontal": 20,
              "paddingVertical": 15,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#545454",
                "fontSize": 14,
              }
            }
          >
            content2
          </Text>
          <View
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "marginTop": 10,
              }
            }
          >
            <View
              style={
                Object {
                  "backgroundColor": "#F6F6F6",
                  "borderRadius": 2,
                  "padding": 8,
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Object {
                    "color": "#747474",
                  }
                }
              >
                5
                 votes
              </Text>
            </View>
            <View
              accessibilityComponentType={undefined}
              accessibilityLabel={undefined}
              accessibilityTraits={undefined}
              accessible={true}
              collapsable={undefined}
              hasTVPreferredFocus={undefined}
              hitSlop={undefined}
              isTVSelectable={true}
              nativeID={undefined}
              onLayout={undefined}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#B02623",
                  "borderRadius": 3,
                  "margin": 5,
                  "opacity": 1,
                  "paddingHorizontal": 20,
                  "paddingVertical": 10,
                }
              }
              testID={undefined}
              tvParallaxProperties={undefined}
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Array [
                    Object {
                      "fontSize": 16,
                      "fontWeight": "bold",
                    },
                    Object {
                      "color": "white",
                    },
                  ]
                }
              >
                UP VOTE
              </Text>
            </View>
            <View
              accessibilityComponentType={undefined}
              accessibilityLabel={undefined}
              accessibilityTraits={undefined}
              accessible={true}
              collapsable={undefined}
              hasTVPreferredFocus={undefined}
              hitSlop={undefined}
              isTVSelectable={true}
              nativeID={undefined}
              onLayout={undefined}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#D8D8D8",
                  "borderRadius": 3,
                  "margin": 5,
                  "opacity": 1,
                  "paddingHorizontal": 20,
                  "paddingVertical": 10,
                }
              }
              testID={undefined}
              tvParallaxProperties={undefined}
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Array [
                    Object {
                      "fontSize": 16,
                      "fontWeight": "bold",
                    },
                    Object {
                      "color": "#000000",
                    },
                  ]
                }
              >
                DOWN VOTE
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`Test HomeScreen component renders correct HomeScreen component with topics 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#E2E2E2",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    data={
      Array [
        Object {
          "content": "content1",
          "id": "123",
          "votes": 1,
        },
        Object {
          "content": "content2",
          "id": "124",
          "votes": 5,
        },
      ]
    }
    disableVirtualization={false}
    extraData={Array []}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "white",
              "elevation": 3,
              "marginVertical": 5,
              "paddingHorizontal": 20,
              "paddingVertical": 15,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#545454",
                "fontSize": 14,
              }
            }
          >
            content1
          </Text>
          <View
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "marginTop": 10,
              }
            }
          >
            <View
              style={
                Object {
                  "backgroundColor": "#F6F6F6",
                  "borderRadius": 2,
                  "padding": 8,
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Object {
                    "color": "#747474",
                  }
                }
              >
                1
                 votes
              </Text>
            </View>
            <View
              accessibilityComponentType={undefined}
              accessibilityLabel={undefined}
              accessibilityTraits={undefined}
              accessible={true}
              collapsable={undefined}
              hasTVPreferredFocus={undefined}
              hitSlop={undefined}
              isTVSelectable={true}
              nativeID={undefined}
              onLayout={undefined}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#B02623",
                  "borderRadius": 3,
                  "margin": 5,
                  "opacity": 1,
                  "paddingHorizontal": 20,
                  "paddingVertical": 10,
                }
              }
              testID={undefined}
              tvParallaxProperties={undefined}
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Array [
                    Object {
                      "fontSize": 16,
                      "fontWeight": "bold",
                    },
                    Object {
                      "color": "white",
                    },
                  ]
                }
              >
                UP VOTE
              </Text>
            </View>
            <View
              accessibilityComponentType={undefined}
              accessibilityLabel={undefined}
              accessibilityTraits={undefined}
              accessible={true}
              collapsable={undefined}
              hasTVPreferredFocus={undefined}
              hitSlop={undefined}
              isTVSelectable={true}
              nativeID={undefined}
              onLayout={undefined}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#D8D8D8",
                  "borderRadius": 3,
                  "margin": 5,
                  "opacity": 1,
                  "paddingHorizontal": 20,
                  "paddingVertical": 10,
                }
              }
              testID={undefined}
              tvParallaxProperties={undefined}
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Array [
                    Object {
                      "fontSize": 16,
                      "fontWeight": "bold",
                    },
                    Object {
                      "color": "#000000",
                    },
                  ]
                }
              >
                DOWN VOTE
              </Text>
            </View>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "white",
              "elevation": 3,
              "marginVertical": 5,
              "paddingHorizontal": 20,
              "paddingVertical": 15,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#545454",
                "fontSize": 14,
              }
            }
          >
            content2
          </Text>
          <View
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "marginTop": 10,
              }
            }
          >
            <View
              style={
                Object {
                  "backgroundColor": "#F6F6F6",
                  "borderRadius": 2,
                  "padding": 8,
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Object {
                    "color": "#747474",
                  }
                }
              >
                5
                 votes
              </Text>
            </View>
            <View
              accessibilityComponentType={undefined}
              accessibilityLabel={undefined}
              accessibilityTraits={undefined}
              accessible={true}
              collapsable={undefined}
              hasTVPreferredFocus={undefined}
              hitSlop={undefined}
              isTVSelectable={true}
              nativeID={undefined}
              onLayout={undefined}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#B02623",
                  "borderRadius": 3,
                  "margin": 5,
                  "opacity": 1,
                  "paddingHorizontal": 20,
                  "paddingVertical": 10,
                }
              }
              testID={undefined}
              tvParallaxProperties={undefined}
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Array [
                    Object {
                      "fontSize": 16,
                      "fontWeight": "bold",
                    },
                    Object {
                      "color": "white",
                    },
                  ]
                }
              >
                UP VOTE
              </Text>
            </View>
            <View
              accessibilityComponentType={undefined}
              accessibilityLabel={undefined}
              accessibilityTraits={undefined}
              accessible={true}
              collapsable={undefined}
              hasTVPreferredFocus={undefined}
              hitSlop={undefined}
              isTVSelectable={true}
              nativeID={undefined}
              onLayout={undefined}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#D8D8D8",
                  "borderRadius": 3,
                  "margin": 5,
                  "opacity": 1,
                  "paddingHorizontal": 20,
                  "paddingVertical": 10,
                }
              }
              testID={undefined}
              tvParallaxProperties={undefined}
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Array [
                    Object {
                      "fontSize": 16,
                      "fontWeight": "bold",
                    },
                    Object {
                      "color": "#000000",
                    },
                  ]
                }
              >
                DOWN VOTE
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
